Mathlib/Data/SetLike/Basic.lean

```diff
@@ -0,0 +1,213 @@
+extensional in this way.
+
+The main use of `SetLike` is for algebraic subobjects (such as
+`Submonoid` and `Submodule`), whose non-proof data consists only of a
+carrier set.  In such a situation, the projection to the carrier set
+is injective.
+
+In general, a type `A` is `SetLike` with elements of type `B` if it
+has an injective map to `Set B`.  This module provides standard
+boilerplate for every `SetLike`: a `coe_sort`, a `coe` to set, a
+`PartialOrder`, and various extensionality and simp lemmas.
+
+A typical subobject should be declared as:
+```
+structure MySubobject (X : Type*) [ObjectTypeclass X] :=
+(carrier : Set X)
+(op_mem' : ∀ {x : X}, x ∈ carrier → sorry ∈ carrier)
+
+namespace MySubobject
+
+variables {X : Type*} [ObjectTypeclass X] {x : X}
+
+instance : SetLike (MySubobject X) X :=
+⟨MySubobject.carrier, λ p q h, by cases p; cases q; congr'⟩
+
+@[simp] lemma mem_carrier {p : MySubobject X} : x ∈ p.carrier ↔ x ∈ (p : Set X) := Iff.rfl
+
+@[ext] theorem ext {p q : MySubobject X} (h : ∀ x, x ∈ p ↔ x ∈ q) : p = q := SetLike.ext h
+
+/-- Copy of a `MySubobject` with a new `carrier` equal to the old one. Useful to fix definitional
+equalities. See Note [range copy pattern]. -/
+protected def copy (p : MySubobject X) (s : Set X) (hs : s = ↑p) : MySubobject X :=
+{ carrier := s,
+  op_mem' := hs.symm ▸ p.op_mem' }
+
+@[simp] lemma coe_copy (p : MySubobject X) (s : Set X) (hs : s = ↑p) :
+  (p.copy s hs : Set X) = s := rfl
+
+lemma copy_eq (p : MySubobject X) (s : Set X) (hs : s = ↑p) : p.copy s hs = p :=
+SetLike.coe_injective hs
+
+end MySubobject
+```
+
+An alternative to `SetLike` could have been an extensional `Membership` typeclass:
+```
+class ExtMembership (α : out_param $ Type u) (β : Type v) extends Membership α β :=
+(ext_iff : ∀ {s t : β}, s = t ↔ ∀ (x : α), x ∈ s ↔ x ∈ t)
+```
+While this is equivalent, `SetLike` conveniently uses a carrier set projection directly.
+
+## Tags
+
+subobjects
+-/
+
+
+/-- A class to indicate that there is a canonical injection between `A` and `Set B`.
+
+This has the effect of giving terms of `A` elements of type `B` (through a `Membership`
+instance) and a compatible coercion to `Type*` as a subtype.
+
+Note: if `SetLike.coe` is a projection, implementers should create a simp lemma such as
+```
+@[simp] lemma mem_carrier {p : MySubobject X} : x ∈ p.carrier ↔ x ∈ (p : Set X) := Iff.rfl
+```
+to normalize terms.
+-/
+class SetLike (A : Type _) (B : outParam <| Type _) where
+  /-- The coercion from a term of a `SetLike` to its corresponding `Set`. -/
+  protected coe : A → Set B
+  /-- The coercion from a term of a `SetLike` to its corresponding `Set` is injective. -/
+  protected coe_injective' : Function.Injective coe
+#align set_like SetLike
+
+attribute [coe] SetLike.coe
+namespace SetLike
+
+variable {A : Type _} {B : Type _} [i : SetLike A B]
+
+-- porting note: is this the right `Coe` class? Or should it be `CoeHead` or `CoeTail`
+instance : CoeTC A (Set B) where coe := SetLike.coe
```

Comment:
`CoeTC` is fine.  This can be neither `CoeHead` nor `CoeTail`, because `B` doesn't appear on the left-hand side, and `A` doesn't appear on the right-hand side of the coercion, respectively.

The linter should complain if change it to `CoeHead` or `CoeTail` :smile: